pkg/usage/metrics.go

```diff
@@ -0,0 +1,208 @@
+package usage
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	authv1 "k8s.io/api/authorization/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+)
+
+type Metrics struct {
+	pageViewsTotal       *prometheus.CounterVec
+	pageImpressionsTotal *prometheus.CounterVec
+	users                *prometheus.GaugeVec
+}
+
+func (m *Metrics) GetCollectors() []prometheus.Collector {
+	return []prometheus.Collector{
+		m.pageViewsTotal,
+		m.pageImpressionsTotal,
+		m.users,
+	}
+}
+
+func (m *Metrics) HandlePageView(perspective string) error {
+	counter, err := m.pageViewsTotal.GetMetricWithLabelValues(perspective)
+	if counter != nil && err == nil {
+		counter.Inc()
+	}
+	return err
+}
+
+func (m *Metrics) HandlePageImpression(perspective string) error {
+	counter, err := m.pageImpressionsTotal.GetMetricWithLabelValues(perspective)
+	if counter != nil && err == nil {
+		counter.Inc()
+	}
+	return err
+}
+
+func (m *Metrics) MonitorUsers(
+	userSettingsClient *http.Client,
+	userSettingsEndpoint string,
+	serviceAccountToken string,
+) {
+	go func() {
+		time.Sleep(3 * time.Second)
+		go m.updateUsersMetric(userSettingsClient, userSettingsEndpoint, serviceAccountToken)
+	}()
+
+	ticker := time.NewTicker(1 * time.Minute) // TODO: Change to every 12 hours?
+	quit := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				m.updateUsersMetric(userSettingsClient, userSettingsEndpoint, serviceAccountToken)
+			case <-quit:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+}
+
+func (m *Metrics) updateUsersMetric(
+	userSettingsClient *http.Client,
+	userSettingsEndpoint string,
+	serviceAccountToken string,
+) error {
+	klog.Info("usage.Metrics: Count console users...\n")
```

Comment:
The update runs now when the console starts and then every 6 hours. Added this to `pkg/usage/metrics.go`:

```go
// We don't expect users to change all the time. Because we check the permission for
// each console user we update these metrics only a few times a day.
// When the bridge starts and then every 6 hours.
const updateConsoleUsersInterval = 6 * time.Hour

// Reduce load peaks when checking permissions for hunders of console users.
// This distributes the requests to 10-20 per second.
// In other words the update takes round about 10 seconds per 100 users.
// Setting this to zero in unit tests.
var delayBetweenConsoleUserPermissionChecks = 100 * time.Millisecond
```